Description: Adapts the region capplet and the language chooser in the user accounts capplet
Forwarded: https://bugzilla.gnome.org/695939, https://bugzilla.gnome.org/695940
Author: Gunnar Hjalmarsson <gunnarhj@ubuntu.com>
Last-Update: 2013-03-21

 Some background that explains the need for this patch:

 * Ubuntu isn't shipped with all languages installed and all locales generated.
   Instead the users install the needed language pack(s), and the corresponding
   locales are generated.

 * The code in accountsservice for handling languages has been extended via Ubuntu
   specific patches. That code is shared by multiple packages. Currently those
   packages are:
   - gnome-control-center
   - language-selector (Xubuntu and Lubuntu will keep using the language-selector
     UI for now)
   - lightdm (for the language chooser in lightdm-gtk-greeter)

 With this patch applied, the g-c-c region capplet and the language chooser in the
 user accounts capplet behave approximately the same way as language-selector. No
 migration code needed when standard Ubuntu drops the language-selector UI.

 Compared with how the region capplet in g-c-c works out of the box, this patch
 results in:

 * Languages are dealt with using 'll' and 'll_CC' language codes instead of
   complete locale names.

 * Selected language is stored in both LANGUAGE and LANG.

 * Only one language list that shows the translations belonging to the installed
   language packs (not a 'short' and a 'long' list).

 * User level settings of language and regional formats are stored as
   accountsservice properties and in ~/.pam_environment.

 * All the formats related locale categories set, not just a subset.
 
 The following changes now live in gnome-desktop3

 * Translation @variants, e.g. ca@valencia, are displayed properly.

 * Locales stored with codeset '.UTF-8' instead of '.utf8'.

 * The 'common_name' field taken into account when parsing the iso-codes XML files
   for language names.

From c62853699f05c91a8de9e430a26b1268a3436dda Mon Sep 17 00:00:00 2001
From: Tim Lunn <tim@feathertop.org>
Date: Sun, 25 Jan 2015 12:42:40 +1100
Subject: [PATCH] ubuntu region language

---
 panels/common/cc-common-language.c   | 100 +++++++++++++++++++++++++--
 panels/common/cc-common-language.h   |   3 +
 panels/region/cc-format-chooser.c    |   2 +
 panels/region/cc-region-panel.c      | 129 +++++++++++++++++++++++++++--------
 panels/user-accounts/um-user-panel.c |   7 +-
 5 files changed, 205 insertions(+), 36 deletions(-)

diff --git a/panels/common/cc-common-language.c b/panels/common/cc-common-language.c
index 9d68ef2..c24d26a 100644
--- a/panels/common/cc-common-language.c
+++ b/panels/common/cc-common-language.c
@@ -167,6 +167,66 @@ cc_common_language_get_current_language (void)
         return language;
 }
 
+gchar *
+cc_common_language_get_property (const gchar *prop_name)
+{
+  GDBusConnection  *bus;
+  gchar            *user_path;
+  GError           *error = NULL;
+  GVariant         *properties;
+  GVariantIter     *iter;
+  gchar            *key;
+  GVariant         *value;
+  gchar            *ret = NULL;
+
+  if (g_strcmp0 (prop_name, "Language") != 0 && g_strcmp0 (prop_name, "FormatsLocale") != 0) {
+    g_warning ("Invalid argument: '%s'", prop_name);
+    return ret;
+  }
+
+  bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, NULL);
+  user_path = g_strdup_printf ("/org/freedesktop/Accounts/User%i", getuid ());
+
+  properties = g_dbus_connection_call_sync (bus,
+                                            "org.freedesktop.Accounts",
+                                            user_path,
+                                            "org.freedesktop.DBus.Properties",
+                                            "GetAll",
+                                            g_variant_new ("(s)", "org.freedesktop.Accounts.User"),
+                                            G_VARIANT_TYPE ("(a{sv})"),
+                                            G_DBUS_CALL_FLAGS_NONE,
+                                            -1,
+                                            NULL,
+                                            &error);
+  if (!properties) {
+    g_warning ("Error calling GetAll() when retrieving properties for %s: %s", user_path, error->message);
+    g_error_free (error);
+                /* g_hash_table_lookup() is not NULL-safe, so don't return NULL */
+                if (g_strcmp0 (prop_name, "Language") == 0)
+                        ret = g_strdup ("en");
+                else
+                        ret = g_strdup ("en_US.UTF-8");
+    goto out;
+  }
+
+  g_variant_get (properties, "(a{sv})", &iter);
+  while (g_variant_iter_loop (iter, "{&sv}", &key, &value)) {
+    if (g_strcmp0 (key, prop_name) == 0) {
+      g_variant_get (value, "s", &ret);
+      break;
+    }
+  }
+
+  g_variant_unref (properties);
+  g_variant_iter_free (iter);
+
+out:
+  g_object_unref (bus);
+  g_free (user_path);
+
+  return ret;
+}
+
 static char *
 get_lang_for_user_object_path (const char *path)
 {
@@ -241,9 +301,18 @@ GHashTable *
 cc_common_language_get_initial_languages (void)
 {
         GHashTable *ht;
+        gchar **langs;
+        gint i;
 
         ht = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
 
+        langs = gnome_get_all_locales ();
+        for (i = 0; langs[i]; i++) {
+                char *language = strtok (langs[i], ".");
+                insert_language (ht, language);
+        }
+        g_strfreev (langs);
+/*
         insert_language (ht, "en_US.UTF-8");
         insert_language (ht, "en_GB.UTF-8");
         insert_language (ht, "de_DE.UTF-8");
@@ -253,10 +322,32 @@ cc_common_language_get_initial_languages (void)
         insert_language (ht, "ja_JP.UTF-8");
         insert_language (ht, "ru_RU.UTF-8");
         insert_language (ht, "ar_EG.UTF-8");
-
+*/
         return ht;
 }
 
+void
+cc_common_language_get_locale (const gchar *language, gchar **locale){
+        gchar *command;
+        GError *error = NULL;
+        /* Get locale that corresponds to the language */
+        command = g_strconcat ("/usr/share/language-tools/language2locale ", language, NULL);
+        if (!g_spawn_command_line_sync (command, locale, NULL, NULL, &error)) {
+               g_warning ("Couldn't get LANG locale: %s", error->message);
+                g_error_free (error);
+                g_free (command);
+
+                return;
+        }
+
+        g_free (command);
+        g_strchomp (*locale);
+        if (strlen (*locale) == 0) {
+                g_warning ("Couldn't get LANG locale -- Copying interrupted");
+                return;
+        }
+
+}
 static void
 foreach_user_lang_cb (gpointer key,
                       gpointer value,
@@ -282,13 +373,14 @@ cc_common_language_add_user_languages (GtkTreeModel *model)
         GtkListStore *store = GTK_LIST_STORE (model);
         GHashTable *user_langs;
         const char *display;
+        const char *lang;
 
         gtk_list_store_clear (store);
 
         user_langs = cc_common_language_get_initial_languages ();
 
         /* Add the current locale first */
-        name = cc_common_language_get_current_language ();
+        lang = cc_common_language_get_property ("Language");
         display = g_hash_table_lookup (user_langs, name);
         if (!display) {
                 char *language = NULL;
@@ -319,8 +411,8 @@ cc_common_language_add_user_languages (GtkTreeModel *model)
         g_hash_table_foreach (user_langs, (GHFunc) foreach_user_lang_cb, store);
 
         /* And now the "Other…" selection */
-        gtk_list_store_append (store, &iter);
-        gtk_list_store_set (store, &iter, LOCALE_COL, NULL, DISPLAY_LOCALE_COL, _("Other…"), -1);
+        //gtk_list_store_append (store, &iter);
+        //gtk_list_store_set (store, &iter, LOCALE_COL, NULL, DISPLAY_LOCALE_COL, _("Other…"), -1);
 
         g_hash_table_destroy (user_langs);
 }
diff --git a/panels/common/cc-common-language.h b/panels/common/cc-common-language.h
index 5571475..b7acaed 100644
--- a/panels/common/cc-common-language.h
+++ b/panels/common/cc-common-language.h
@@ -44,6 +44,7 @@ guint    cc_common_language_add_available_languages (GtkListStore     *store,
                                                      GHashTable       *user_langs);
 gboolean cc_common_language_has_font                (const gchar  *locale);
 gchar   *cc_common_language_get_current_language    (void);
+gchar   *cc_common_language_get_property            (const gchar  *prop_name);
 
 GHashTable *cc_common_language_get_initial_languages   (void);
 GHashTable *cc_common_language_get_user_languages      (void);
@@ -56,6 +57,8 @@ void     cc_common_language_select_current_language (GtkTreeView  *treeview);
 
 void     cc_common_language_add_user_languages (GtkTreeModel *model);
 
+void    cc_common_language_get_locale (const gchar *language, gchar **locale);
+
 G_END_DECLS
 
 #endif
diff --git a/panels/region/cc-format-chooser.c b/panels/region/cc-format-chooser.c
index 5944781..b2ccd0f 100644
--- a/panels/region/cc-format-chooser.c
+++ b/panels/region/cc-format-chooser.c
@@ -167,7 +167,9 @@ set_locale_id (GtkDialog   *chooser,
 
                 } else {
                         /* mark as unselected */
+
                         gtk_widget_set_opacity (check, 0.0);
+
                 }
         }
         g_list_free (children);
diff --git a/panels/region/cc-region-panel.c b/panels/region/cc-region-panel.c
index 4c20f6e..ed1f7ae 100644
--- a/panels/region/cc-region-panel.c
+++ b/panels/region/cc-region-panel.c
@@ -406,6 +406,47 @@ set_system_region (CcRegionPanel *self,
 }
 
 static void
+set_formats_locale (const gchar *formats_locale)
+{
+        GDBusProxy  *proxy;
+        GError      *error = NULL;
+        gchar       *user_path;
+        GVariant    *ret;
+
+        user_path = g_strdup_printf ("/org/freedesktop/Accounts/User%i", getuid ());
+        proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
+                                               G_DBUS_PROXY_FLAGS_NONE,
+                                               NULL,
+                                               "org.freedesktop.Accounts",
+                                               user_path,
+                                               "org.freedesktop.Accounts.User",
+                                               NULL,
+                                               &error);
+        if (!proxy) {
+                g_warning ("Couldn't get accountsservice proxy for %s: %s", user_path, error->message);
+                g_error_free (error);
+                g_free (user_path);
+                return;
+        }
+
+        ret = g_dbus_proxy_call_sync (proxy,
+                                      "SetFormatsLocale",
+                                      g_variant_new ("(s)", formats_locale),
+                                      G_DBUS_CALL_FLAGS_NONE,
+                                      -1,
+                                      NULL,
+                                      &error);
+        if (!ret) {
+                g_warning ("Couldn't set FormatsLocale: %s", error->message);
+                g_error_free (error);
+        } else
+                g_variant_unref (ret);
+
+        g_object_unref (proxy);
+        g_free (user_path);
+}
+
+static void
 update_region (CcRegionPanel *self,
                const gchar   *region)
 {
@@ -416,6 +457,7 @@ update_region (CcRegionPanel *self,
         } else {
                 if (g_strcmp0 (region, priv->region) == 0)
                         return;
+                set_formats_locale(region);
                 g_settings_set_string (priv->locale_settings, KEY_REGION, region);
                 if (priv->login_auto_apply)
                         set_system_region (self, region);
@@ -577,7 +619,7 @@ update_region_from_setting (CcRegionPanel *self)
         CcRegionPanelPrivate *priv = self->priv;
 
         g_free (priv->region);
-        priv->region = g_settings_get_string (priv->locale_settings, KEY_REGION);
+        priv->region = cc_common_language_get_property ("FormatsLocale");
         update_region_label (self);
 }
 
@@ -610,15 +652,17 @@ static void
 update_language_from_user (CcRegionPanel *self)
 {
 	CcRegionPanelPrivate *priv = self->priv;
-        const gchar *language;
+        const gchar *language, *locale;
 
-        if (act_user_is_loaded (priv->user))
+        if (act_user_is_loaded (priv->user)){
                 language = act_user_get_language (priv->user);
+                cc_common_language_get_locale (language, &locale);
+        }
         else
-                language = setlocale (LC_MESSAGES, NULL);
+                locale = setlocale (LC_MESSAGES, NULL);
 
         g_free (priv->language);
-        priv->language = g_strdup (language);
+        priv->language = g_strdup (locale);
         update_language_label (self);
 }
 
@@ -1414,6 +1458,16 @@ setup_input_section (CcRegionPanel *self)
         update_buttons (self);
 }
 
+static gchar *
+strip_quotes (const gchar *str)
+{
+        if ((g_str_has_prefix (str, "\"") && g_str_has_suffix (str, "\""))
+          || (g_str_has_prefix (str, "'") && g_str_has_suffix (str, "'")))
+                return g_strndup (str + 1, strlen (str) - 2);
+        else
+                return g_strdup (str);
+}
+
 static void
 on_localed_properties_changed (GDBusProxy     *proxy,
                                GVariant       *changed_properties,
@@ -1428,34 +1482,50 @@ on_localed_properties_changed (GDBusProxy     *proxy,
                 const gchar **strv;
                 gsize len;
                 gint i;
-                const gchar *lang, *messages, *time;
+                gchar *lang, *language, *messages, *time;
 
                 strv = g_variant_get_strv (v, &len);
 
-                lang = messages = time = NULL;
+                lang = language = messages = time = NULL;
                 for (i = 0; strv[i]; i++) {
                         if (g_str_has_prefix (strv[i], "LANG=")) {
-                                lang = strv[i] + strlen ("LANG=");
+                                lang = strip_quotes (strv[i] + strlen ("LANG="));
+                        } else if (g_str_has_prefix (strv[i], "LANGUAGE=")) {
+                                gchar *tmp = strip_quotes (strv[i] + strlen ("LANGUAGE="));
+                                gchar **tokens = g_strsplit (tmp, ":", 2);
+                                language = g_strdup (tokens[0]);
+                                g_free (tmp);
+                                g_strfreev (tokens);
                         } else if (g_str_has_prefix (strv[i], "LC_MESSAGES=")) {
-                                messages = strv[i] + strlen ("LC_MESSAGES=");
+                                messages = strip_quotes (strv[i] + strlen ("LC_MESSAGES="));
                         } else if (g_str_has_prefix (strv[i], "LC_TIME=")) {
-                                time = strv[i] + strlen ("LC_TIME=");
+                                time = strip_quotes (strv[i] + strlen ("LC_TIME="));
                         }
                 }
+                g_free (strv);
+
                 if (!lang) {
-                        lang = setlocale (LC_MESSAGES, NULL);
+                        lang = g_strdup ("en_US.UTF-8");
                 }
-                if (!messages) {
-                        messages = lang;
+
+                if (!language) {
+                        if (messages)
+                                language = g_strdup (messages);
+                        else
+                                language = g_strdup (lang);
                 }
                 if (!time) {
-                        time = lang;
+                        time = g_strdup (lang);
                 }
                 g_free (priv->system_language);
-                priv->system_language = g_strdup (messages);
+                priv->system_language = g_strdup (language);
                 g_free (priv->system_region);
                 priv->system_region = g_strdup (time);
                 g_variant_unref (v);
+                g_free (lang);
+                g_free (language);
+                g_free (messages);
+                g_free (time);
 
                 update_language_label (self);
         }
@@ -1525,29 +1595,28 @@ set_localed_locale (CcRegionPanel *self)
 	CcRegionPanelPrivate *priv = self->priv;
         GVariantBuilder *b;
         gchar *s;
+        gchar *lang;
+        gint i;
+
+        cc_common_language_get_locale (priv->system_language, &lang);
 
         b = g_variant_builder_new (G_VARIANT_TYPE ("as"));
-        s = g_strconcat ("LANG=", priv->system_language, NULL);
+        s = g_strconcat ("LANG=", lang, NULL);
         g_variant_builder_add (b, "s", s);
         g_free (s);
 
-        if (g_strcmp0 (priv->system_language, priv->system_region) != 0) {
-                s = g_strconcat ("LC_TIME=", priv->system_region, NULL);
-                g_variant_builder_add (b, "s", s);
-                g_free (s);
-                s = g_strconcat ("LC_NUMERIC=", priv->system_region, NULL);
-                g_variant_builder_add (b, "s", s);
-                g_free (s);
-                s = g_strconcat ("LC_MONETARY=", priv->system_region, NULL);
-                g_variant_builder_add (b, "s", s);
-                g_free (s);
-                s = g_strconcat ("LC_MEASUREMENT=", priv->system_region, NULL);
-                g_variant_builder_add (b, "s", s);
-                g_free (s);
-                s = g_strconcat ("LC_PAPER=", priv->system_region, NULL);
+        s = g_strconcat ("LANGUAGE=", priv->system_language, NULL);
+        g_variant_builder_add (b, "s", s);
+        g_free (s);
+        const gchar *format_categories[] = { "LC_NUMERIC", "LC_TIME",
+           "LC_MONETARY", "LC_PAPER", "LC_IDENTIFICATION", "LC_NAME",
+           "LC_ADDRESS", "LC_TELEPHONE", "LC_MEASUREMENT", NULL };
+        for (i = 0; format_categories[i] != NULL; i++) {
+                s = g_strconcat (format_categories[i], "=", priv->system_region, NULL);
                 g_variant_builder_add (b, "s", s);
                 g_free (s);
         }
+
         g_dbus_proxy_call (priv->localed,
                            "SetLocale",
                            g_variant_new ("(asb)", b, TRUE),
diff --git a/panels/user-accounts/um-user-panel.c b/panels/user-accounts/um-user-panel.c
index e7c4a21..fd7e050 100644
--- a/panels/user-accounts/um-user-panel.c
+++ b/panels/user-accounts/um-user-panel.c
@@ -905,8 +905,9 @@ show_user (ActUser *user, CcUserPanelPrivate *d)
 
         name = NULL;
         lang = g_strdup (act_user_get_language (user));
+        cc_common_language_get_locale (lang, &lang);
         if ((!lang || *lang == '\0') && act_user_get_uid (user) == getuid ()) {
-                lang = cc_common_language_get_current_language ();
+                lang = cc_common_language_get_property ("Language");
                 act_user_set_language (user, lang);
         }
 
@@ -1086,6 +1087,7 @@ language_response (GtkDialog         *dialog,
         ActUser *user;
         const gchar *lang, *account_language;
         gchar *current_language;
+        gchar *locale;
         gchar *name = NULL;
         gboolean self_selected;
 
@@ -1099,8 +1101,9 @@ language_response (GtkDialog         *dialog,
         self_selected = act_user_get_uid (user) == geteuid ();
 
         lang = cc_language_chooser_get_language (GTK_WIDGET (dialog));
+        cc_common_language_get_locale(act_user_get_language (user),&locale);
         if (lang) {
-                if (g_strcmp0 (lang, account_language) != 0) {
+                if (g_strcmp0 (lang, locale) != 0) {
                         act_user_set_language (user, lang);
 
                         /* Do not show the notification if the locale is already used. */
-- 
2.1.4


